src/libsync/vfs/cfapi/cfapiwrapper.cpp

```diff
@@ -309,8 +325,148 @@ OCC::Optional<OCC::PinStateEnums::PinState> OCC::CfApiWrapper::PlaceHolderInfo::
+    LocalFree(stringSid);
+    return QString::fromWCharArray(stringSid);
+}
+
+std::unique_ptr<TOKEN_USER> getCurrentTokenInformation()
+{
+    const auto tokenHandle = GetCurrentThreadEffectiveToken();
+
+    auto tokenInfoSize = DWORD{0};
+
+    const auto tokenSizeCallSucceeded = ::GetTokenInformation(tokenHandle, TokenUser, nullptr, 0, &tokenInfoSize);
+    const auto lastError = GetLastError();
+    Q_ASSERT(!tokenSizeCallSucceeded && lastError == ERROR_INSUFFICIENT_BUFFER);
+    if (tokenSizeCallSucceeded || lastError != ERROR_INSUFFICIENT_BUFFER) {
+        qCCritical(lcCfApiWrapper) << "GetTokenInformation for token size has failed with error" << lastError;
+        return std::unique_ptr<TOKEN_USER>();
+    }
+
+    std::unique_ptr<TOKEN_USER> tokenInfo;
+
+    tokenInfo.reset(reinterpret_cast<TOKEN_USER*>(new char[tokenInfoSize]));
+    if (!::GetTokenInformation(tokenHandle, TokenUser, tokenInfo.get(), tokenInfoSize, &tokenInfoSize)) {
+        qCCritical(lcCfApiWrapper) << "GetTokenInformation failed with error" << lastError;
+        return std::unique_ptr<TOKEN_USER>();
+    }
+
+    return tokenInfo;
+}
+
+QString retrieveWindowsSID()
+{
+    if (const auto tokenInfo = getCurrentTokenInformation()) {
+        return convertSidToStringSid(tokenInfo->User.Sid);
+    }
+
+    return QString();
+}
+
+bool createSyncRootRegistryKeys(const QString &providerName, const QString &folderAlias, const QString &displayName, const QString &accountDisplayName, const QString &syncRootPath)
+{
+    // We must set specific Registry keys to make the progress bar refresh correctly and also add status icons into Windows Explorer
+    // More about this here: https://docs.microsoft.com/en-us/windows/win32/shell/integrate-cloud-storage
+    const auto windowsSID = retrieveWindowsSID();
+    Q_ASSERT(!windowsSID.isEmpty());
+    if (windowsSID.isEmpty()) {
+        qCWarning(lcCfApiWrapper) << "Failed to set Registry keys for shell integration, as windowsSID is empty. Progress bar will not work.";
+        return false;
+    }
+
+    // syncRootId should be: [storage provider ID]![Windows SID]![Account ID]![FolderAlias] (FolderAlias is a custom part added here to be able to register multiple sync folders for the same account)
+    // folder registry keys go like: Nextcloud!S-1-5-21-2096452760-2617351404-2281157308-1001!user@nextcloud.lan:8080!0, Nextcloud!S-1-5-21-2096452760-2617351404-2281157308-1001!user@nextcloud.lan:8080!1, etc. for each sync folder
+    const auto syncRootId = QString("%1!%2!%3!%4").arg(providerName).arg(windowsSID).arg(accountDisplayName).arg(folderAlias);
+
+    const QString providerSyncRootIdRegistryKey = QString() % R"(SOFTWARE\Microsoft\Windows\CurrentVersion\Explorer\SyncRootManager\)" % syncRootId;
+    const QString providerSyncRootIdUserSyncRootsRegistryKey = QString() % providerSyncRootIdRegistryKey % R"(\UserSyncRoots\)";
+
+    const QVector<std::tuple<HKEY, QString, QString, int, QVariant>> registryKeysToSet = {
```

Comment:
I'd declare a local struct type instead of relying on tuple. This way you'd get properly named members and wouldn't need the call to tie. Here we pack it all to unpack right way in the loop.

Reply:
@er-vin I thought using tuple makes sense here, instead of creating structs every time we need a construct like this vector with multiple fields? Otherwise, is it forbidden to use tuples at all? I am surprised a bit. What is the downside of using tuples? I thought it is a convenient feature of the STL after-all? Could you elaborate (maybe a link to some article with a detailed explanation)?

Reply:
@er-vin I've discussed this with the guys :) Everyone seems to like the struct :).